app/api_topology.go

```diff
@@ -144,11 +147,13 @@ func handleWebsocket(
 		re, err := rep.Report(ctx, reportTimestamp)
 		if err != nil {
 			log.Errorf("Error generating report: %v", err)
+			span.Finish()
```

Comment:
Can't we defer `span.Finish()` at the top instead of adding it before every `return`?

Reply:
I actually coded that first time, but it doesn't work.  `defer` runs things at the end of the function, and we want to run at the end of the loop.
Possibly the whole thing would be neater if the loop body was written as a separate function, which could then use `defer` as you suggest.

Reply:
> Possibly the whole thing would be neater if the loop body was written as a separate function, which could then use defer as you suggest.

Not sure if that would make it better or worse, your call :)

Reply:
I refactored the code to move several variables into a struct and the loop body into a member function on that struct.  See what you think.